### PR TITLE
sql: enforce minimum privileges for "admin" role.

### PR DESCRIPTION
**Pull request by mberhault**
This fixes an issue caught during QA: #21872

Also changed:
* validate used to enforce "no more privileges than root", this was
incorrect and should just be "must have no more than allowed privileges"
* switched to using the "upgrade descriptor" migration helper
* dropped the "multiple allowed privileges" in favor of just one

Release note (bug fix): enforce minimum privileges for "admin" role.

### CODE REVIEW

File: pkg/sqlmigrations/migrations.go [hunk 1]
```diff
@@ -929,27 +887,49 @@ func upgradeTableDescsWithFn(
...
+							// upgrades every last old-format table descriptor, including those
+							// that are dropping. Otherwise, the user could upgrade to a version
+							// without support for reading the old format before the drop
+							// completes, leaving a broken table descriptor and the table's
+							// remaining data around forever. This isn't just a theoretical
+							// concern: consider that dropping a large table can take several
+							// days, while upgrading to a new version can take as little as a
+							// few minutes.
+							table.UpVersion = true
+							if err := table.Validate(ctx, txn); err != nil {
```
Comment (mberhault) [inline pkg/sqlmigrations/migrations.go]: The `table.Validate` call is new. We really need to do that before writing a descriptor.

File: pkg/sqlmigrations/migrations.go [hunk 2]
```diff
@@ -885,20 +826,37 @@ func addRootToAdminRole(ctx context.Context, r runner) error {
...
+	return upgradeDescsWithFn(ctx, r, tableDescFn, nil)
+}
+
+// ensureAdminRolePrivileges ensures that all descriptors have privileges
+// for the admin role.
+//
+// TODO(mberhault): Remove this migration in v2.1.
+func ensureAdminRolePrivileges(ctx context.Context, r runner) error {
+	tableDescFn := func(desc *sqlbase.TableDescriptor) (bool, error) {
+		return desc.Privileges.MaybeAddAdminPrivileges(desc.ID), nil
```
Comment (mberhault) [inline pkg/sqlmigrations/migrations.go]: This is technically taken care of by `grantAdminPrivileges` before this change, and by `upgradeTableDescsToInterleavedFormatVersion`, but it doesn't hurt to specify it again.

File: pkg/sql/logictest/testdata/logic_test/system [hunk 3]
```diff
@@ -376,42 +375,77 @@ GRANT GRANT, SELECT ON system.namespace TO testuser
...
+REVOKE GRANT ON DATABASE system FROM admin
+
+statement error user admin must have exactly {GRANT, SELECT} privileges on system object with ID=.*
+REVOKE GRANT ON system.namespace FROM admin
+
+statement error user admin does not have privileges
+REVOKE ALL ON system.namespace FROM admin
+
 # Some tables (we test system.lease here) allow multiple privilege sets for
-# backwards compatibility, but still enforce that no user has more privileges
```
Comment (mberhault) [inline pkg/sql/logictest/testdata/logic_test/system]: this was an oversight on my part when reviewing the changes to this code before (when "allowedPrivileges" became a list). It was never meant to check that "users have no more privileges than root", it was supposed to be "users have allowed privileges". The confusion came from the fact that "root" had exactly the allowed privileges, so it was a shortcut to not have to convert types again (from privilege kind to privilege bits).

**Comment by benesch** [comment]
Looks good! But one big caveat:

> this was an oversight on my part when reviewing the changes to this code before (when "allowedPrivileges" became a list). It was never meant to check that "users have no more privileges than root", it was supposed to be "users have allowed privileges". The confusion came from the fact that "root" had exactly the allowed privileges, so it was a shortcut to not have to convert types again (from privilege kind to privilege bits).

The fact that root sometimes has create/drop privileges on system.settings, system.lease, system.eventlog, system.rangelog, and system.ui is a big mistake. I think non-root users should be limited to the semi-safe "desired" permissions.

---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/system, line 431 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7eW3JGFwjQjbCtitt6:-L7eW3JGFwjQjbCtitt7:b-vveebw) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/logictest/testdata/logic_test/system#L431)):*
> ```
> # Superusers must have exactly one of the described privilege sets.
> statement ok
> GRANT ALL ON system.lease TO testuser
> ```

This doesn't seem right? testuser shouldn't be able to drop system.lease. (Neither should root, but that ship has sailed.)

---

*[pkg/sql/sqlbase/privilege.go, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7edyh17lUq_w0YeHg9:-L7edyh17lUq_w0YeHgA:b-fdqhvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/privilege.go#L226)):*
> ```Go
> 	if err := p.validateRequiredSuperuser(id, allowedPrivileges, AdminRole); err != nil {
> 		return err
> 	}
> ```

The admin role should only ever have the "desired" permissions. It should be an error if you try to grant ALL on system.lease to the admin role, for example.

---

*[pkg/sql/sqlbase/privilege.go, line 238 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7ed5u08A6BqmfqkMcZ:-L7ed5u08A6BqmfqkMc_:bdqtyhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/privilege.go#L238)):*
> ```Go
> 		// privilege is not allowed by any item in the list.
> 		var userError error
> 		for _, allowedPriv := range allowedPrivileges {
> ```

Likewise, I don't think we should allow users besides root to have anything more the "desired" privileges on a table. 

---

*[pkg/sql/sqlbase/structured.go, line 487 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7ebf-ePzClON3UFEyQ:-L7ebf-ePzClON3UFEyR:bb2a2x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/structured.go#L487)):*
> ```Go
> // maybeUpgradeFormatVersion transforms the TableDescriptor to the latest
> // FormatVersion (if it's not already there) and returns true if any changes
> // were made.
> ```

Recommend adding a warning like "Don't call directly; call MaybeFillInDescriptor`.

---

*[pkg/sql/sqlbase/structured.go, line 1050 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7ebbem4urzYDWBRX6f:-L7ebben87fCDaJiAFNv:b-u6gwu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/structured.go#L1050)):*
> ```Go
> 	// We maintain forward compatibility, so if you see this error message with a
> 	// version older that what this client supports, then there's a
> 	// maybeUpgradeFormatVersion missing from some codepath.
> ```

s/maybeUpgradeFormatVersion/MaybeFillInDescriptor`

---

*[pkg/sql/tests/system_table_test.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7ebRM7bmb1iWiKP2Uj:-L7ebRM8ken9AwkqEslN:b-a8bo5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/tests/system_table_test.go#L139)):*
> ```Go
> 			t.Fatalf("test: %+v, err: %v", test, err)
> 		}
> 		// Privileges for the admin user get automatically added in GetTable. We need to remove them
> ```

`GetTable`? Do you mean `CreateTestTableDescriptor`?

---

*[pkg/sqlmigrations/migrations.go, line 849 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-L7eXLkMc2g6iAOcd3Xz:-L7eXLkMc2g6iAOcd3Y-:b91573c) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sqlmigrations/migrations.go#L849)):*
> ```Go
> }
> 
> var upgradeTableDescBatchSize int64 = 50
> ```

Optional nit: if you go through CI again, this should be called `upgradeDescBatchSize` now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23876#-:-L7edLOw_DKnxfENHnVJ:b7nfew6)*

**Comment by mberhault** [comment]
I agree, but that mistake has been there from the beginning. The stricted default privileges were added later, but we had to keep allowing the wider `ALL` allowed privileges to not break existing deployments. This hasn't changed since last time (the only thing that changed was requiring that users can't have more privileges than root which was a mistake in the PR that introduced the privilege list). We could probably fix the actual problem by adding a migration to update any existing tables to the stricter default privileges, and we probably should. It's technically a separate issue though.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/system, line 431 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7eW3JGFwjQjbCtitt6:-L7ehJJ-254cSq_wwN0G:b-vf4329) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/logictest/testdata/logic_test/system#L431)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This doesn't seem right? testuser shouldn't be able to drop system.lease. (Neither should root, but that ship has sailed.)
</blockquote></details>

This is the "multiple allowed privilege sets" issue. The change here only removed the "user can't have more privileges than root". Even before that, you could always grant everything to root then grant everything to a user.

---

*[pkg/sql/sqlbase/privilege.go, line 226 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7edyh17lUq_w0YeHg9:-L7ehrKD67pTzg_fSzAn:bl9dsmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/privilege.go#L226)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

The admin role should only ever have the "desired" permissions. It should be an error if you try to grant ALL on system.lease to the admin role, for example.
</blockquote></details>

I can restrict the allowed privileges for `admin` since it's new (and was migrated with the stricter privileges). I think this would be better addressed by the proper fix though.

---

*[pkg/sql/sqlbase/privilege.go, line 238 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7ed5u08A6BqmfqkMcZ:-L7ei7D51iyKhXJNFLoy:brft27k) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/privilege.go#L238)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Likewise, I don't think we should allow users besides root to have anything more the "desired" privileges on a table. 
</blockquote></details>

Just like root, we can't do much here without potentially impacting exiting deployment (where someone granted wide privileges to a user). 

---

*[pkg/sql/sqlbase/structured.go, line 487 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7ebf-ePzClON3UFEyQ:-L7eiVIJ8BKELPJInoZo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/structured.go#L487)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Recommend adding a warning like "Don't call directly; call MaybeFillInDescriptor`.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.go, line 1050 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7ebbem4urzYDWBRX6f:-L7eibFl1nPR1fh7-NKv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/sqlbase/structured.go#L1050)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

s/maybeUpgradeFormatVersion/MaybeFillInDescriptor`
</blockquote></details>

Done.

---

*[pkg/sql/tests/system_table_test.go, line 139 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7ebRM7bmb1iWiKP2Uj:-L7eis0qBT05zn30Be59:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sql/tests/system_table_test.go#L139)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

`GetTable`? Do you mean `CreateTestTableDescriptor`?
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 849 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7eXLkMc2g6iAOcd3Xz:-L7ejOsK2X-EHwQLt0dW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69bd4e1058ca80df49e0bc48ecd31d7fab0c9a22/pkg/sqlmigrations/migrations.go#L849)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Optional nit: if you go through CI again, this should be called `upgradeDescBatchSize` now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-:-L7egiUQ7AAXY7TSntTq:bpassz7)*

**Comment by mberhault** [comment]
New commit removes the list of allowed privileges, rolling this into a migration + descriptor modification at load time. We now always ensure that "root" and "admin" both have exactly the required privileges and that other users do not exceed the allowed privileges. The migration will need to be repeated for 2.2 (it has a comment to that effect).

---

Review status: 3 of 14 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-:-L7ezvnF1fVjI85sKHyz:b1v6xz2)*

**Comment by benesch** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r2, 10 of 10 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/sqlbase/privilege_test.go, line 392 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7f_Z3z2Pgr5NpF6YVY:-L7f_Z3z2Pgr5NpF6YVZ:bnpkm40) ([raw file](https://github.com/cockroachdb/cockroach/blob/f442ae4f45b78008e0ad73b9c3ef03ac66b20327/pkg/sql/sqlbase/privilege_test.go#L392)):*
> ```Go
> 				AdminRole:         systemPrivs,
> 				"foo":             privilege.List{},
> 				"bar":             privilege.List{},
> ```

Ideally "bar" would have `privilege.UPDATE` and `privilege.SELECT`, and the test would verify that the valid priv (`SELECT`) remains.

---

*[pkg/sql/sqlbase/system.go, line 233 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7fYoaF1PNdQaGc2Gob:-L7fYoaF1PNdQaGc2Goc:bijj7ew) ([raw file](https://github.com/cockroachdb/cockroach/blob/f442ae4f45b78008e0ad73b9c3ef03ac66b20327/pkg/sql/sqlbase/system.go#L233)):*
> ```Go
> // where id is guaranteed to exist.
> func SystemDesiredPrivileges(id ID) privilege.List {
> 	return SystemAllowedPrivileges[id]
> ```

This method might as well be deleted now in favor of just inlining the map access, but don't go back through CI just for this.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-:-L7famJu1TOZhccf926t:bnfp4nl)*

**Comment by mberhault** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sqlbase/privilege_test.go, line 392 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7f_Z3z2Pgr5NpF6YVY:-L7fcRsoCGL2GNkHY4PV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f442ae4f45b78008e0ad73b9c3ef03ac66b20327/pkg/sql/sqlbase/privilege_test.go#L392)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ideally "bar" would have `privilege.UPDATE` and `privilege.SELECT`, and the test would verify that the valid priv (`SELECT`) remains.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/system.go, line 233 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-L7fYoaF1PNdQaGc2Gob:-L7fd8GTBAEsEPdVenlE:bmq4zz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f442ae4f45b78008e0ad73b9c3ef03ac66b20327/pkg/sql/sqlbase/system.go#L233)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This method might as well be deleted now in favor of just inlining the map access, but don't go back through CI just for this.
</blockquote></details>

Done. I need to squash my commits anyway, and this touches four files, so I'm cool running CI again.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876)*

**Comment by mberhault** [comment]
TFTQR (Q = Quick)

**Comment by benesch** [comment]
:shipit: thanks for doing the hard work!

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23876#-:-L7fdsws51I_mnTA_55P:bv2c5r3)*

**Pull request merged**